ledger/block/src/header/metadata/mod.rs

```diff
@@ -80,40 +81,50 @@ impl<N: Network> Metadata<N> {
-    pub fn is_valid(&self) -> bool {
-        match self.height == 0u32 {
-            true => self.is_genesis(),
-            false => {
-                // Ensure the network ID is correct.
-                self.network == N::ID
-                    // Ensure the round is nonzero.
-                    && self.round != 0u64
-                    // Ensure the height is nonzero.
-                    && self.height != 0u32
-                    // Ensure the round is at least as large as the height.
-                    && self.round >= self.height as u64
-                    // Ensure the coinbase target is at or above the minimum.
-                    && self.coinbase_target >= N::GENESIS_COINBASE_TARGET
-                    // Ensure the proof target is at or above the minimum.
-                    && self.proof_target >= N::GENESIS_PROOF_TARGET
-                    // Ensure the coinbase target is larger than the proof target.
-                    && self.coinbase_target > self.proof_target
-                    // Ensure the last coinbase target is at or above the minimum.
-                    && self.last_coinbase_target >= N::GENESIS_COINBASE_TARGET
-                    // Ensure the last coinbase timestamp is after the genesis timestamp.
-                    && self.last_coinbase_timestamp >= N::GENESIS_TIMESTAMP
-                    // Ensure the timestamp in the block is after the genesis timestamp.
-                    && self.timestamp > N::GENESIS_TIMESTAMP
+    pub fn check_validity(&self) -> Result<()> {
+        if self.height == 0u32 {
+            // [`Self::is_genesis`] performs its own validity checks.
+            if !self.is_genesis().with_context(|| "Genesis block check failed")? {
+                bail!("Block at height 0 is not a genesis block");
             }
+            return Ok(());
+        }
+
+        // Ensure the network ID is correct.
+        ensure!(self.network == N::ID, "Invalid network ID");
+        ensure!(self.round > 0u64, "Invalid round");
+        ensure!(self.round >= self.height as u64, "Round must be greater or equal to height");
+        ensure!(self.proof_target >= N::GENESIS_PROOF_TARGET, "Invalid proof target");
+
+        ensure!(self.last_coinbase_timestamp >= N::GENESIS_TIMESTAMP, "Ensure last coinbase timestamp");
+        ensure!(self.timestamp > N::GENESIS_TIMESTAMP, "Invalid timeestamp");
+
+        if self.coinbase_target < N::GENESIS_COINBASE_TARGET {
+            bail!(
+                "Invalid coinbase target: Was {actual} but expected {min} or greater.",
+                actual = self.coinbase_target,
+                min = N::GENESIS_COINBASE_TARGET,
+            );
+        }
+
+        if self.proof_target < N::GENESIS_PROOF_TARGET {
+            bail!(
+                "Invalid proof target: Was {actual} but expected {min} or greater.",
+                actual = self.proof_target,
+                min = N::GENESIS_PROOF_TARGET,
+            );
```

Comment:
Why not stick to the ensure! and >/>= convention? Makes it more consistent and easier to read.